osu.Game.Tests/Visual/SongSelect/TestSceneBeatmapRecommendations.cs

```diff
@@ -0,0 +1,168 @@
+                                _ => 0
+                            };
+
+                            userRequest.TriggerSuccess(new User
+                            {
+                                Username = @"Dummy",
+                                Id = 1001,
+                                Statistics = new UserStatistics
+                                {
+                                    PP = pp
+                                }
+                            });
+                            break;
+                    }
+                };
+                // Force recommender to calculate its star ratings again
+                recommender.APIStateChanged(API, APIState.Online);
+            });
+        }
+
+        [Test]
+        public void TestPresentedBeatmapIsRecommended()
+        {
+            var importFunctions = new List<Func<BeatmapSetInfo>>();
+
+            for (int i = 0; i < 5; i++)
+            {
+                importFunctions.Add(importBeatmap(i, new List<RulesetInfo> { null, null, null, null, null }));
+            }
+
+            for (int i = 0; i < 5; i++)
+            {
+                presentAndConfirm(importFunctions[i], i, 2);
+            }
+        }
+
+        [Test]
+        public void TestBestRulesetIsRecommended()
+        {
+            var osuRuleset = rulesets.AvailableRulesets.First(r => r.ID == 0);
+            var taikoRuleset = rulesets.AvailableRulesets.First(r => r.ID == 1);
+            var catchRuleset = rulesets.AvailableRulesets.First(r => r.ID == 2);
+            var maniaRuleset = rulesets.AvailableRulesets.First(r => r.ID == 3);
+
+            var osuImport = importBeatmap(0, new List<RulesetInfo> { osuRuleset });
+            var mixedImport = importBeatmap(1, new List<RulesetInfo> { taikoRuleset, catchRuleset, maniaRuleset });
+
+            // Make sure we are on standard ruleset
+            presentAndConfirm(osuImport, 0, 1);
+
+            // Present mixed difficulty set, expect ruleset with highest star difficulty
+            presentAndConfirm(mixedImport, 1, 3);
+        }
+
+        [Test]
+        public void TestSecondBestRulesetIsRecommended()
+        {
+            var osuRuleset = rulesets.AvailableRulesets.First(r => r.ID == 0);
+            var taikoRuleset = rulesets.AvailableRulesets.First(r => r.ID == 1);
+            var catchRuleset = rulesets.AvailableRulesets.First(r => r.ID == 2);
+
+            var osuImport = importBeatmap(0, new List<RulesetInfo> { osuRuleset });
+            var mixedImport = importBeatmap(1, new List<RulesetInfo> { taikoRuleset, catchRuleset, taikoRuleset });
+
+            // Make sure we are on standard ruleset
+            presentAndConfirm(osuImport, 0, 1);
+
+            // Present mixed difficulty set, expect ruleset with highest star difficulty
+            presentAndConfirm(mixedImport, 1, 2);
+        }
+
+        private Func<BeatmapSetInfo> importBeatmap(int importID, List<RulesetInfo> rulesets)
```

Comment:
Not sure I like the way this is written too much. A little complicated to parse but I'm not sure if it's just me.

Would probably pre-construct the beatmap info in the first step of each test, store them to a variable/field and then operate on the created info through `Func`s in the asserts (so pass the info instead of passing magic indices to `presentAndConfirm`).

Reply:
Seems like you resolved this yourself in 2913a8183538f823858aace50834f9e38a74b726 but I think that code was copied from [TestScenePresentBeatmap](https://github.com/ppy/osu/blob/93269606cd78c1193d561ba649f254afa1255c27/osu.Game.Tests/Visual/Navigation/TestScenePresentBeatmap.cs#L88-L129). Maybe something over there needs to be changed too

Reply:
separate PR even if so.